### ISSUE TITLE
Bad connection driver error

### ISSUE THREAD

**Issue by collisonchris** [johnnadratowski/golang-neo4j-bolt-driver#57]
I've been using the driver in an application for about 5 months and have had very few issues. However, recently I've noticed I've been getting `bad connection` errors. It always seems to be pop up from the decoder `read()` call. The error and stack trace look like this:
**Error Message**
`Couldn't read expected bytes for message length. Read: 0 Expected: 2.
Internal Error(*errors.errorString):driver: bad connection`
**Stack Trace**
```
Stack Trace:

goroutine 159 [running]:
runtime/debug.Stack(0x5557536b466f, 0x46, 0xc4204c54d0)
	/usr/lib/go/src/runtime/debug/stack.go:24 +0xa9
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/errors.Wrap(0x555753ad5120, 0xc4202204a0, 0x5557536b466f, 0x46, 0xc4204c54d0, 0x1, 0x1, 0xc42139f420)
	/go/src/<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/errors/errors.go:38 +0x18a
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/encoding.Decoder.read(0x555753ad5440, 0xc42139f420, 0xc4201461c0, 0x5557531abada, 0x70, 0x555753a73b20)
	/go/src/<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/encoding/decoder.go:45 +0x420
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/encoding.Decoder.Decode(0x555753ad5440, 0xc42139f420, 0xc4201461c0, 0xc42139f420, 0xc420146150, 0xffff, 0x555753a40d60)
	/go/src/<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/encoding/decoder.go:93 +0x45
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver.(*boltConn).consume(0xc42139f420, 0x5557536b731a, 0x4b8, 0xc4262329f0, 0x0)
	/go/src/<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/conn.go:550 +0xb4
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver.(*boltConn).sendRunPullAllConsumeAll(0xc42139f420, 0x5557536b731a, 0x4b8, 0xc4262329f0, 0x5557531abbbc, 0x480, 0x8, 0x10, 0xc42055e6b0, 0x0, ...)
	/go/src/<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver/conn.go:697 +0x70
<my_package_structure>/vendor/github.com/johnnadratowski/golang-neo4j-bolt-driver.(*boltStmt).ExecNeo(0xc42091ee40, 0xc4262329f0, 0x55575369a780, 0xf, 0xc4205fa898, 0x0)
```

I am running in the pooled connection mode with 5 connections currently. The code below is how I am building the pool and retrieving a connection object from the pool.

```
var boltConnectionString = "<sanitized_connection_string>"

type NeoClient struct {
	Pool bolt.DriverPool
}

func NewNeo4jClient(hostname string, port string, user string, password string) (NeoClient, error) {
	pool, err := bolt.NewDriverPool(boltConnectionString, 5)
	boltLogs.SetLevel("error")
	return NeoClient{pool}, err
}

func getConnectionFromPool(pool bolt.DriverPool) bolt.Conn {
	conn, err := pool.OpenPool()
	if (err != nil) {
		platform.Log().WithError(err).Error("Unable to get connection from Pool!")
	}
	return conn
}
```

The errors seem to happen in very short bursts and are consistently come in groups of 5 and always very close together(usually same millisecond). In most ~15k message batches, I'll see this error exactly 10 times. This makes me wonder if this is some retry/reconnect logic kicking in after 5 failures? I don't see any errors on the Neo4j server logs side about closing connection, connection reset by peer etc. 

Any ideas as to what's going on here? Are there any configuration changes I could make that might help with this?

**Comment by markuszm** [comment]
I am also getting these type of errors. 
What type of queries/statements do you execute to trigger these driver errors?

In my case it happens with very long running statements (a call to apoc.refactor.mergeNodes) and I believe the connection times out as it is set at 1 Minute by default. Still, after such an error I get these bad driver errors for the next queries too. I am not sure if it is also a problem with the server hanging with heavy calculations as I am running neo4j locally in a Docker container on my dual-core laptop.

**Comment by collisonchris** [comment]
Our scenario is quite different. We're ingesting messages off a message broker and using a persistent connection pool to send a merge to our Neo4j instance. Our Neo4j instance is a VM and our app that uses the driver is in a docker container in a Kubernetes pod. Most of the time it's fine and processes messages smoothly but after long layoffs these messages tend to pop up. I've been digging into the Bolt protocol details here: https://boltprotocol.org/v1/#examples

Basically it looks like at some point during query issue the connection has gone stale or fails in some way. Line 697 in `conn.go` is where the errors begin:

```
func (c *boltConn) sendRunPullAllConsumeAll(query string, args map[string]interface{}) (interface{}, interface{}, []interface{}, error) {
	err := c.sendRunPullAll(query, args)
	if err != nil {
		return nil, nil, nil, err
	}

	runSuccess, err := c.consume()//LINE 697
	if err != nil {
		return runSuccess, nil, nil, err
	}

	records, pullSuccess, err := c.consumeAll()
	return runSuccess, pullSuccess, records, err
}
```

If it made it that far it's already sent the `RUN` and the `PULLALL` messages and gotten responses successfully in the `sendRunPullAll` call, it appears to only fail when it goes to consume in the `RECORD` step from the server. 

I'm gonna turn on `INFO` level logging in our environment and see if it gives any more details.

**Comment by collisonchris** [comment]
After digging into this more it appears that connections do become stale after a sufficient idle period. I came up with a couple of different ideas how to deal with the issue. The first is to check the to see if the underlying connection can be read from before returning it. The existing code was doing a nil check on the connection before returning it which wouldn't detect `io.EOF` errors or timed out connections.
```
func connectionNilOrClosed(conn *boltConn) (bool) {
	if(conn.conn == nil) {//nil check before attempting read
		return true
	}
	conn.conn.SetReadDeadline(time.Now())
	zero := make ([]byte, 0)
	_, err := conn.conn.Read(zero)//read zero bytes to validate connection is still alive
	if err != nil {
		log.Error("Bad Connection state detected", err)//the error caught here could be a io.EOF or a timeout, either way we want to log the error & return true
		return true
	}
	return false
}
```


This should take care of the the `driver: bad connection` errors we've been seeing. PR for this change is here:  #59

If this isn't successful I will try to downcast the connection to a `net.TCPConn` and turn on keep alive behavior. I tinkered with this locally but it didn't seem to work as I expected; may be worth following up on if this doesn't work.

**Comment by bluestreek18** [comment]
I'm still having the same issue after this change.

` Error: An error occurred getting result of exec command: <nil>
        Couldn't read expected bytes for message length. Read: 0 Expected: 2.
Internal Error(*errors.errorString):driver: bad connection`

**Comment by collisonchris** [comment]
Yeah, this doesn't seem to fix the issue in our environment either. Believe it's a sync/cross-threading issue, I've been able to reproduce this issue with this simple POC: https://github.com/collisonchris/neo4j-test

**Comment by bluestreek18** [comment]
I would think with OpenPool() you should be ok since you should receive a new connection each call to OpenPool() even across goroutines. In your example you are using one bolt connection per goroutine and closing it after. Perhaps OpenPool() can return the same connection multiple times if you call it too quickly?
